staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/implementation.go

```diff
@@ -0,0 +1,314 @@
+		}
+		evalResult, _, err := compilationResult.Program.Eval(va)
+		if err != nil {
+			policyDecision.Kind = Deny
+			policyDecision.Message = fmt.Sprintf("failed rule: %v, with evaluation error: %v, for object: %s", v.policy.Spec.Validations[i].Expression, err, va.object)
+		} else if evalResult != types.True {
+			policyDecision.Kind = Deny
+			// FIXME: properly format the error message.
+			policyDecision.Message = fmt.Sprintf("failed rule: %v", v.policy.Spec.Validations[i].Expression)
+		} else {
+			policyDecision.Kind = Admit
+		}
+	}
+
+	return decisions, nil
+}
+
+// From vendor/k8s.io/apiserver/pkg/endpoints/openapi/openapi.go
+func friendlyName(name string) string {
+	nameParts := strings.Split(name, "/")
+	// Reverse first part. e.g., io.k8s... instead of k8s.io...
+	if len(nameParts) > 0 && strings.Contains(nameParts[0], ".") {
+		parts := strings.Split(nameParts[0], ".")
+		for i, j := 0, len(parts)-1; i < j; i, j = i+1, j-1 {
+			parts[i], parts[j] = parts[j], parts[i]
+		}
+		nameParts[0] = strings.Join(parts, ".")
+	}
+	return strings.Join(nameParts, ".")
+}
+
+func createAdmissionRequest(attr admission.Attributes) *admissionv1.AdmissionRequest {
+	uid := api_types.UID(uuid.NewUUID())
+
+	gvk := attr.GetKind()
+	gvr := attr.GetResource()
+	subresource := attr.GetSubresource()
+
+	// FIXME: how to get resource GVK, GVR and subresource?
+	requestGVK := attr.GetKind()
+	requestGVR := attr.GetResource()
+	requestSubResource := attr.GetSubresource()
+
+	aUserInfo := attr.GetUserInfo()
+	var userInfo authenticationv1.UserInfo
+	if aUserInfo != nil {
+		userInfo = authenticationv1.UserInfo{
+			Extra:    make(map[string]authenticationv1.ExtraValue),
+			Groups:   aUserInfo.GetGroups(),
+			UID:      aUserInfo.GetUID(),
+			Username: aUserInfo.GetName(),
+		}
+		// Convert the extra information in the user object
+		for key, val := range aUserInfo.GetExtra() {
+			userInfo.Extra[key] = authenticationv1.ExtraValue(val)
+		}
+	}
+
+	dryRun := attr.IsDryRun()
+
+	return &admissionv1.AdmissionRequest{
+		UID: uid,
+		Kind: metav1.GroupVersionKind{
+			Group:   gvk.Group,
+			Kind:    gvk.Kind,
+			Version: gvk.Version,
+		},
+		Resource: metav1.GroupVersionResource{
+			Group:    gvr.Group,
+			Resource: gvr.Resource,
+			Version:  gvr.Version,
+		},
+		SubResource: subresource,
+		RequestKind: &metav1.GroupVersionKind{
+			Group:   requestGVK.Group,
+			Kind:    requestGVK.Kind,
+			Version: requestGVK.Version,
+		},
+		RequestResource: &metav1.GroupVersionResource{
+			Group:    requestGVR.Group,
+			Resource: requestGVR.Resource,
+			Version:  requestGVR.Version,
+		},
+		RequestSubResource: requestSubResource,
+		Name:               attr.GetName(),
+		Namespace:          attr.GetNamespace(),
+		Operation:          admissionv1.Operation(attr.GetOperation()),
+		UserInfo:           userInfo,
+		Object: runtime.RawExtension{
+			Object: attr.GetObject(),
+		},
+		OldObject: runtime.RawExtension{
+			Object: attr.GetOldObject(),
+		},
```

Comment:
Leave Object and OldObject unset since we don't provide access to them via `request` (we want developers to access them through the root variables).